src/main/com/yetanalytics/lrs/pedestal/interceptor/xapi/statements/attachment.cljc

```diff
@@ -133,145 +122,110 @@
+          {}
+          multiparts))
 
-(defn validate-statements-multiparts
-  "Validate and return statements and their multipart attachments"
-  [statements multiparts]
-  (let [;; collect the attachments per statement and reduce over them
-        {valid-statements    :s-acc
-         valid-multiparts    :a-acc
-         leftover-multiparts :mps}
+(defn validate-multiparts
+  "Given a list of statements and a list of multiparts, return valid multiparts,
+  deduplicated."
+  [statements
+   multiparts]
+  (let [{:keys [sha2s
+                mpart-map]}
         (reduce
-         (fn [{:keys [mps] :as m}
-              [s & att-objs]]
-           ;; match the attachment objects to a multipart
-           ;; or assert that they are a fileURL
-           (let [;; Reduce over multiparts and find relevant attachments
-                 [next-mps
-                  valid-matched]
-                 (reduce
-                  (fn [[mps' acc :as state] ao]
-                    (if-let [[sha mp :as match]
-                             (make-sha-multipart-pair
-                              s mps' ao)]
-                      [(if (< 1 (count (get mps' sha)))
-                         (update mps' sha #(into [] (rest %)))
-                         (dissoc mps' sha))
-                       (conj acc match)]
-                      state))
-                  [mps []]
-                  att-objs)
-
-                 valid-shas    (map first valid-matched)]
-             (-> m
-                 (update :s-acc conj s)
-                 (update :a-acc into (map second valid-matched))
-                 (assoc :mps
-                        next-mps))))
-         {;; accumulators for the statements + attachments
-          :s-acc []
-          :a-acc []
-          ;; the multiparts to join up, by Hash
-          :mps   (multipart-map multiparts)}
-         (statements-attachments statements))]
-    (if (seq leftover-multiparts)
-      ;; If we have leftovers, it's bad
+         (fn [{:keys [sha2s mpart-map]
+               :as state}
+              {:keys [statement
+                      attachment-path]
+               {:strs [sha2 fileUrl]
+                :as att-obj} :attachment}]
+           (if-let [match-mp (get mpart-map sha2)]
+             (cond-> (update state :sha2s conj sha2)
+               ;; Validate + recreate sig mp
+               (and (= "attachments"
+                       (first attachment-path))
+                    (sig? att-obj))
+               (update-in
+                [:mpart-map sha2]
+                (partial validate-sig
+                         statement
+                         att-obj)))
+             ;; Allow attachments w/o a fileUrl to silently drop
+             (if fileUrl
+               state
+               (throw
+                ;; TODO: name of this error?
+                (ex-info "Invalid multipart format"
+                         {:type ::invalid-multipart-format})))))
```

Comment:
May as well come up with a proper name while we're at it. In particular, the error message and type should specify that the error is because neither attachment data nor a `fileUrl` exists (which is more specific than the generic "invalid multipart format").

Reply:
I thought this was being referenced by other stuff and that would widen the scope of the PR (since this is what it returned before in the same case). However, that does not appear to be the case, so I'll rename/remessage.

Reply:
https://github.com/yetanalytics/lrs/pull/81/commits/bcda832044eab35fc2d16d5ffc1244657d70042b